apps/s_server.c

```diff
@@ -586,6 +587,56 @@ static int get_ocsp_resp_from_responder(SSL *s, tlsextstatusctx *srctx,
     return ret;
 }
 
+static int get_ocsp_resp_from_responder(SSL *s, tlsextstatusctx *srctx,
+                                        STACK_OF(OCSP_RESPONSE) **sk_resp)
+{
+    X509 *x = NULL;
+    int ret = SSL_TLSEXT_ERR_NOACK;
+    int i, num = 1;
+    STACK_OF(X509) *server_certs = NULL;
+    OCSP_RESPONSE *resp = NULL;
+
+    if (*sk_resp != NULL)
+        sk_OCSP_RESPONSE_pop_free(*sk_resp, OCSP_RESPONSE_free);
+
+    *sk_resp = sk_OCSP_RESPONSE_new_null();
+
+    SSL_get0_chain_certs(s, &server_certs);
+    if (server_certs != NULL && (SSL_version(s) >= TLS1_3_VERSION)) {
+        /* certificate chain is available */
+        num = sk_X509_num(server_certs);
+    }
+    else {
+        /*
+         * certificate chain is not available,
+         * set num to 1 for server certificate
+         */
+        num = 1;
+    }
+
+    sk_OCSP_RESPONSE_reserve(*sk_resp, num);
+
+    for (i = -1; i < num - 1; i++) {
+        if (i == -1) {
+            /* get OCSP response for server certificate first */
+            x = SSL_get_certificate(s);
+        } else {
+            /*
+             * for each certificate in chain (except root) get
+             * the OCSP response
+             */
+            x = sk_X509_value(server_certs, i);
+        }
+
+        resp = NULL;
+        ret = get_ocsp_resp_from_responder_single(s, x, srctx, &resp);
+        if (ret == SSL_TLSEXT_ERR_OK && resp != NULL)
```

Comment:
OCSP stapling appears to be described extremely weakly by RFC 8446, in particular since the obsoleted RFC 6961 was much better here.
Yet I agree that also in case there are gaps, the client must know which OCSP response is for which cert.

Reply:
Again, if the OCSP responses are not read from a file, you get them from the server and then the order is clear - no need to use there the complicated checks (which are inefficient and might go wrong) on the order of the responses.

Reply:
I could simplify the check to 
``` 
if (resp != NULL)
            sk_OCSP_RESPONSE_insert(*sk_resp, resp, -1);
```
would that be ok?
I'm still not happy to relay on the order of responses on the stack and insert gaps to it.
Later I could not distinguish between OCSP responses received from server which are in correct order and responses read from files, which might are not in correct order.
Also in a later stage, when somebody wants to extend the functionality to cache the responses it would be easier not relaying on the order of the requests in the stack.